# Task 4.5- Data Comsistancy Checks

Step 1- Follow procudures outlined in exercise (walkthrough)

Step 2- Complete task

1-Run the df.describe() function on your df_ords dataframe. Using your new knowledge about how to interpret the output of this function, share in a markdown cell whether anything about the data looks off or should be investigated further.

2-Check for mixed-type data in your df_ords dataframe.
If you find mixed-type data, fix it. The column in question should contain observations of a single data type.

3-Run a check for missing values in your df_ords dataframe.

4-In a markdown cell, report your findings and propose an explanation for any missing values you find.
Address the missing values using an appropriate method.

5-In a markdown cell, explain why you used your method of choice.

6-Run a check for duplicate values in your df_ords data.

7-In a markdown cell, report your findings and propose an explanation for any duplicate values you find.
Address the duplicates using an appropriate method.

8-In a markdown cell, explain why you used your method of choice.

Export your final, cleaned df_prods and df_ords data as “.csv” files in your “Prepared Data” folder and give them appropriate, succinct names.

#Importing Libraries

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

#Import Data

In [2]:
df_ords = pd.read_csv(r'C:\Users\kenwh\Instacart Basket Anaylsis\02 Data\Prepared Data\orders_wrangled.csv', index_col = False)

In [3]:
df_prods = pd.read_csv(r'C:\Users\kenwh\Instacart Basket Anaylsis\02 Data\Original Data\products.csv', index_col = False)

# 4.5 Walkthrough

In [4]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

# Create a practice data frame wiht mixed data types

In [5]:
#create a dataframe
df_test = pd.DataFrame()

In [6]:
# create a mixed type column
df_test['mix'] = ['a', 'b', '1', True]

In [7]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [27]:
# check for mixed types
for col in df_test.columns.tolist():
  weird = (df_test[[col]].map(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

In [9]:
df_test['mix'] = df_test['mix'].astype('str')

# Missing Values

In [ ]:
# Find missing values in products dataframe
# The output shows there are 16 missing values in product_name colum

In [10]:
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [11]:
# creating a subset with just the missing values
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [12]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [13]:
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

In [14]:
df_prods.shape

(49693, 5)

In [15]:
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [16]:
df_prods_clean.shape

(49677, 5)

# Duplicate Values

In [17]:
# extract duplicate observations from df_prods_clean and store it in dataframe df_dups
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [18]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [19]:
df_prods_clean.shape

(49677, 5)

In [20]:
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [21]:
df_prods_clean_no_dups.shape

(49672, 5)

# Start of Task 4.5

Question 2- Run the df.describe() function on your df_ords dataframe. Using your new knowledge about how to interpret the output of this function, share in a markdown cell whether anything about the data looks off or should be investigated further.
Tip: Keep an eye on min and max values!

In [23]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

In [24]:
#remove scientific notation- tried out chatgpt to find notation
df_ords.describe().apply(lambda x: round(x, 2))

Unnamed: 0    order_id     user_id  order_number  orders_day_of_week  \
count  3421083.00  3421083.00  3421083.00    3421083.00          3421083.00   
mean   1710541.00  1710542.00   102978.21         17.15                2.78   
std     987581.74   987581.74    59533.72         17.73                2.05   
min          0.00        1.00        1.00          1.00                0.00   
25%     855270.50   855271.50    51394.00          5.00                1.00   
50%    1710541.00  1710542.00   102689.00         11.00                3.00   
75%    2565811.50  2565812.50   154385.00         23.00                5.00   
max    3421082.00  3421083.00   206209.00        100.00                6.00   

       order_hour_of_day  days_since_prior_order  
count         3421083.00              3214874.00  
mean               13.45                   11.11  
std                 4.23                    9.21  
min                 0.00                    0.00  
25%                10.00                    4.00  
50%                13.00                    7.00  
75%                16.00                   15.00  
max                23.00                   30.00

No clear issues with the data in this dataframe, the mins and maxs are all within the expected ranges, and the count is the same accross each column. 

Question 3- Check for mixed-type data in your df_ords dataframe.

In [29]:
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].map(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

Since it returned no results there are no columns with mixed data

Question 5 - Run a check for missing values in your df_ords dataframe.
In a markdown cell, report your findings and propose an explanation for any missing values you find

In [32]:
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

The days since prior order column has 206209 missing values

In [41]:
df_miss= df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [42]:
df_miss

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0                 0   2539329        1             1                   2   
11               11   2168274        2             1                   2   
26               26   1374495        3             1                   1   
39               39   3343014        4             1                   6   
45               45   2717275        5             1                   3   
...             ...       ...      ...           ...                 ...   
3420930     3420930    969311   206205             1                   4   
3420934     3420934   3189322   206206             1                   3   
3421002     3421002   2166133   206207             1                   6   
3421019     3421019   2227043   206208             1                   1   
3421069     3421069   3154581   206209             1                   3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 7 columns]

In [38]:
df_ords_clean = df_ords.copy()

In [39]:
df_ords_clean['days_since_prior_order'].fillna(0, inplace=True)

In [43]:
df_miss_2= df_ords_clean[df_ords_clean['days_since_prior_order'].isnull() == True]

In [44]:
df_miss_2

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

Given what we know about the column (days since prior order) and the fact the amount of missing orders (206209) exactly matches the amount of uses as seen by the user_id column, I chose to replace the missing values with a zero. This value most likely represents the first order that each person placed. When they first placed an order, they would not have had a prior order to compare to, and thus makes sense that there would not be a value imputed in this column. By changing it to a zero, it still shows that they had zero days since a prior order while also ensuring that all values are filled and it dose not mess up other calculations and tests we can preform on the database.

Just to be safe, I created it as a new databased called df_ords_clean, and kept the other database as df_ords, meaning that if I need t backtrace or go back to having the values as missing I can. 

The test at the end shows that all the missing values in df_ords_clean have indeed been replaced with 0s.


Question 7- Run a check for duplicate values in your df_ords data.
In a markdown cell, report your findings and propose an explanation for any duplicate values you find.

In [45]:
df_ords_duplicates=df_ords_clean[df_ords_clean.duplicated()]

In [46]:
df_ords_duplicates

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

The test above shows that the df_ords_clean dataframe does not have any duplicate data that needs to be handled.

Question 9 - Exporting the df_prods and df_ords_clean to prefared data folder

In [48]:
path = r'C:\Users\kenwh'

In [49]:
path

'C:\\Users\\kenwh'

In [52]:
df_ords_clean.to_csv(os.path.join(path,'Instacart Basket Anaylsis','02 Data','Prepared Data','orders_cleaned.csv'))

In [53]:
df_prods_clean.to_csv(os.path.join(path,'Instacart Basket Anaylsis','02 Data','Prepared Data', 'product_cleaned.csv'))